pkg/resources/management.cattle.io/v3/nodedriver/validator.go

```diff
@@ -0,0 +1,185 @@
+package nodedriver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/webhook/pkg/admission"
+	objectsv3 "github.com/rancher/webhook/pkg/generated/objects/management.cattle.io/v3"
+	admissionv1 "k8s.io/api/admission/v1"
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	gvr = schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodedrivers",
+	}
+
+	driverInUse = &admissionv1.AdmissionResponse{
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "This driver is in use by existing nodes and cannot be disabled",
+		},
+		Allowed: false,
+	}
+)
+
+// Validator ValidatingWebhook for management clusters.
+type Validator struct {
+	admitter admitter
+}
+
+type admitter struct {
+	api k8sAPI
+}
+
+// interface that holds the methods necessary to hit the k8s api and fetch
+// whether rke1/2 resources are deleted
+type k8sAPI interface {
+	rke1ResourcesDeleted(ctx context.Context, driver *v3.NodeDriver) (bool, error)
+	rke2ResourcesDeleted(ctx context.Context, driver *v3.NodeDriver) (bool, error)
+}
+
+type k8sAPIImpl struct {
+	rest rest.Interface
```

Comment:
We don't need to use a raw `rest.Interface` for our lookups. Instead, we should use one of our generated clients and caches. You can generate new ones for RKE1 and RKE2 machines here https://github.com/rancher/webhook/blob/release/v0.3/pkg/codegen/main.go#L30

Reply:
Since these machines are dynamic, we cannot generate new clients for the rke1 or rke2 machines here. You can look here for an example of how we do this in rancher/rancher: https://github.com/rancher/rancher/blob/release/v2.7/pkg/controllers/capr/machineprovision/controller.go#L502

tl;dr we use `unstructured.Unstructured` with a dynamic client.

Reply:
Ahh, my bad I always forget we have those dynamic CRDs. I think we should replicate the same behavior you linked and use [dynamic](https://github.com/rancher/lasso/tree/master/pkg/dynamic) to keep our clients consistent throughout Rancher.

Reply:
Updated to use the `dynamic.Controller` class, injected through the `NewValidator(...)` function.